Creating a script that fetches ropsten contract addresses

dimpar opened the pull request:
Now we can point to the published npm packages of `keep-core`, `keep-ecdsa`, `tbtc` and fetch respective contract adressess. The `update-ropsten-subgraph.js` script updates these addresses in `subgraph.ropsten.yaml` so that it can
be deployed to the subgraph. A short description was added to the README.md

dimpar committed:
Creating a script that fetches ropsten contract addresses

Now we can point to the published npm packages of keep-core, keep-ecdsa
and tbtc and fetch contract adressess. The updated-ropsten-subgraph.js
script updates these addresses in subgraph.ropsten.yaml so that it can
be deployed to the subgraph. Updated README.md

michalinacienciala reviewed (approved):
Looking good to me. Leaving for @nkuba to review as well.

nkuba commented on update-ropsten-subgraph.js:
```diff
@@ -0,0 +1,189 @@
+const { resolve } = require("path")
+
+const contractOwnerProvider = new HDWalletProvider(
+  process.env.CONTRACT_OWNER_ETH_ACCOUNT_PRIVATE_KEY,
```
Is this provider is used only to read data from the chain or is it expected to submit transactions?
If it's read-only we don't need to pass the private key, just `const web3 = new Web3(process.env.ETH_RPC_URL)`.

dimpar replied:
Good point, we need read-only. fdd6802

nkuba commented on update-ropsten-subgraph.js:
```diff
@@ -0,0 +1,189 @@
+  const TBTCDepositToken = {
+    'address': await getDeploymentContractAddress(TBTCDepositTokenArtifact, TBTCDepositTokenContractName),
+    'startBlock': await getDeploymentBlockNumber(TBTCDepositTokenArtifact, TBTCDepositTokenContractName)
+  }
```
Could we simplify the code by combining three functions into one function?

```suggestion
  async function getContractData(packageName, contractName) {
    let artifactRaw
    try {
      artifactRaw = readFileSync(
        resolve(`./node_modules/${packageName}/artifacts/${contractName}.json`)
      )
    } catch (err) {
      throw new Error(`failed to read artifact file: ${err.message}`)
    }

    let artifact
    try {
      artifact = JSON.parse(artifactRaw)
    } catch (err) {
      throw new Error(`failed to parse artifact file: ${err.message}`)
    }

    if (!artifact.networks[networkID]) {
      throw new Error(
        `configuration for network ${networkID} not found in ${contractName}`
      )
    }

    if (!artifact.networks[networkID].address) {
      throw new Error(
        `missing address for network ${networkID} in ${contractName}`
      )
    }

    if (!artifact.networks[networkID].transactionHash) {
      throw new Error(
        `missing transactionHash for network ${networkID} in ${contractName}`
      )
    }

    const deployTransaction = await web3.eth.getTransaction(
      artifact.networks[networkID].transactionHash
    )

    return {
      address: artifact.networks[networkID].address,
      startBlock: deployTransaction.blockNumber,
    }
  }

  // tbtc contracts
  const TBTCDepositToken = await getContractData(
    keepTbtcPackageName,
    "TBTCDepositToken"
  )
```

dimpar replied:
Sure we can: f71ed35

nkuba commented on package.json:
```diff
@@ -15,9 +15,9 @@
-    "@keep-network/keep-core": "^1.4.1",
-    "@keep-network/keep-ecdsa": "^1.3.1",
-    "@keep-network/tbtc": "^1.1.0",
+    "@keep-network/keep-core": "^1.8.0-ropsten.3",
+    "@keep-network/keep-ecdsa": "^1.7.0-ropsten.2",
+    "@keep-network/tbtc": "^1.1.2-ropsten.3",
```
So once we merge this PR to main the package dependencies will be updated from mainnet to ropsten. I prefer this repository to be able to support multiple networks. Not sure yet what would be the best approach though.

dimpar replied:
This change is actually for local development only and when the workflow runs these versions [will be replaced](https://github.com/keep-network/allthekeeps/blob/allthekeeps-on-ropsten/.github/workflows/allthekeeps-testnet.yml#L54-L56). The whole [issue](https://keepnetwork.atlassian.net/browse/CP-39) of ATK PRs was to make it work on Ropsten since the mainnet is supported by the community. But yea, being able to switch is certainly a good option in case we'd want to switch our test network (although I guess that would require much more coordination). I think what we can do without digging too much is to use the same approach as we have in workflow that runs ATK, ie. passing `keep-*` package versions `KEEP_CORE_VERSION`, `KEEP_ECDSA_VERSION` & `KEEP_TBTC_VERSION`. As for now if you don't pass specific package versions, it will just get the latest Ropsten packages (this also answers your [other comment](https://github.com/keep-network/keep-subgraph/pull/1#discussion_r702965373)). I can revert this change, but then it will work as follows: if we pass the versions, then the project will work with those packages; otherwise, it will point to mainnet. wdyt?

nkuba commented on README.md:
```diff
@@ -18,3 +18,11 @@ You may also find my companion graph [keep-pricefeed](https://github.com/miracle
+# Development Ropsten (Auto)
+- Run `CONTRACT_OWNER_ETH_ACCOUNT_PRIVATE_KEY=<key> ETH_RPC_URL=<url> SUBGRAPH_DEPLOY_KEY=<slug> GRAPH_DEPLOY_KEY=<key> ./deploy-subgraph.sh`
+
+  - `CONTRACT_OWNER_ETH_ACCOUNT_PRIVATE_KEY`: for HDWalletProvider
```
This is probably not necessary, please see the comment in `update-ropsten-subgraph.js`.

dimpar replied:
4a6cb93

nkuba commented on deploy-subgraph.sh:
```diff
@@ -0,0 +1,55 @@
+yarn global add @graphprotocol/graph-cli
+
+# for CI purposes, it will get the latest deployed npm packages for the current CI jobs
+if [[ $KEEP_CORE_VERSION != "" ]] && [[ $KEEP_ECDSA_VERSION != "" ]] && [[ $KEEP_TBTC_VERSION != "" ]]
```
What will happen if any of these variables is empty?

dimpar replied:
It will take the Ropsten versions, but please see this [comment](https://github.com/keep-network/keep-subgraph/pull/1#discussion_r706135328)

dimpar committed:
Using web3 with rpc_url only for read-only mode

dimpar committed:
Simplifying reading of the deployed artifacts

dimpar committed:
Passing a private key is not needed

dimpar committed:
Locking graph-cli global version

dimpar committed:
Locking graph-cli global version 0.21.1

dimpar committed:
Echoing subgraph api

nkuba commented on deploy-subgraph.sh:
```diff
@@ -0,0 +1,57 @@
+# for CI purposes, it will get the latest deployed npm packages for the current CI jobs
+if [[ $KEEP_CORE_VERSION != "" ]] && [[ $KEEP_ECDSA_VERSION != "" ]] && [[ $KEEP_TBTC_VERSION != "" ]]
+then
+    yarn upgrade \
+      @keep-network/keep-core@$KEEP_CORE_VERSION  \
+      @keep-network/keep-ecdsa@$KEEP_ECDSA_VERSION \
+      @keep-network/tbtc@$KEEP_TBTC_VERSION  --exact
```
```suggestion
    yarn upgrade --exact \
      @keep-network/keep-core@$KEEP_CORE_VERSION  \
      @keep-network/keep-ecdsa@$KEEP_ECDSA_VERSION \
      @keep-network/tbtc@$KEEP_TBTC_VERSION
```

dimpar replied:
Done in 010084b

nkuba commented on update-ropsten-subgraph.js:
```diff
@@ -0,0 +1,123 @@
+const Web3 = require("web3")
+const HDWalletProvider = require("@truffle/hdwallet-provider")
```
We don't use it anymore.

```suggestion
```

dimpar replied:
Done in 010084b

nkuba commented on update-ropsten-subgraph.js:
```diff
@@ -0,0 +1,123 @@
+    const deployTransaction = await web3.eth.getTransaction(
+      artifact.networks[networkID].transactionHash
+    )
+
```
Let's check if `deployTransaction.blockNumber` exists and throw an error otherwise.

dimpar replied:
Done in 010084b

nkuba commented on update-ropsten-subgraph.js:
```diff
@@ -0,0 +1,123 @@
+      )
+    }
+
+    const deployTransaction = await web3.eth.getTransaction(
+      artifact.networks[networkID].transactionHash
+    )
```
Maybe it would be worth wrapping it with try/catch as well?

dimpar replied:
Done in 010084b

dimpar committed:
Improvements of ropsten subgraph

- Minor change how we upgrade @keep-network dependency projects
- Adding "if" guards when something goes wrong
